storage-proofs-update/src/gadgets.rs

```diff
@@ -0,0 +1,604 @@
+
+    // The number of digests generated per partition.
+    let digests_per_partition = (challenges as f32 / challenges_per_digest as f32).ceil() as u64;
+
+    let mut generated_bits = Vec::with_capacity(challenges);
+
+    for j in 0..digests_per_partition {
+        // The index of the current digest across all partition proofs:
+        // `digest_index = k * digests_per_partition + j`.
+        let digest_index =
+            AllocatedNum::alloc(cs.namespace(|| format!("digest_index_{}", j)), || {
+                let k = partition
+                    .get_value()
+                    .ok_or(SynthesisError::AssignmentMissing)?;
+                let digest_index = k * Fr::from(digests_per_partition) + Fr::from(j);
+                Ok(digest_index)
+            })?;
+
+        // `digests_per_partition` and `j` are (unallocated) constants.
+        cs.enforce(
+            || format!("digest_index_{} == k * digests_per_partition + {}", j, j),
+            |lc| {
+                lc + (Fr::from(digests_per_partition), partition.get_variable())
+                    + (Fr::from(j), CS::one())
+            },
+            |lc| lc + CS::one(),
+            |lc| lc + digest_index.get_variable(),
+        );
+
+        // `digest = H(comm_r_new || digest_index)`
+        let digest = <PoseidonHasher as Hasher>::Function::hash2_circuit(
+            cs.namespace(|| format!("digest_{}", j)),
+            &comm_r_new,
+            &digest_index,
+        )?;
+
+        // Allocate `digest` as `Fr::NUM_BITS` bits.
+        let digest_bits = allocated_num_to_allocated_bits(
+            cs.namespace(|| format!("digest_{}_bits", j)),
+            &digest,
+        )?;
+
+        // We may not take all available challenge bits from the last digest.
+        let challenges_to_take = if j == digests_per_partition - 1 {
+            challenges - generated_bits.len()
+        } else {
+            challenges_per_digest
+        };
+
+        digest_bits
+            .chunks(bits_per_challenge)
+            .take(challenges_to_take)
+            .for_each(|bits| {
+                generated_bits.push(bits.to_vec());
+            });
+    }
+
+    Ok(generated_bits)
+}
+
+pub fn get_challenge_high_bits<CS: ConstraintSystem<Fr>>(
+    mut cs: CS,
+    // little-endian
+    c_generated_bits: &[AllocatedBit],
+    h_select_bits: &[AllocatedBit],
+    hs: &[usize],
+) -> Result<AllocatedNum<Fr>, SynthesisError> {
+    assert_eq!(h_select_bits.len(), hs.len());
+
+    let bit_len = c_generated_bits.len();
+
+    let c_generated_bits: Vec<Boolean> = c_generated_bits.iter().cloned().map(Into::into).collect();
+
+    // For each `h in hs`, get the `h` high bits of the challenge's randomly generated bits (i.e.
+    // the challenge's bits sans the partition-index). Scale each "high" value by the corresponding
+    // bit of `h_select` producing a vector containing `hs.len() - 1` zeros and 1 "high" value
+    // selected via `h_select_bits`.
```

Comment:
`(i.e. the challenge's bits sans the partition-index).` this part of the comment is wrong, I think it got out of date.